test/testinfra.py

```diff
@@ -225,28 +235,51 @@ def request(self, method, resource, data="", headers=None, return_headers_on_unm
                 if connected:
                     raise
                 self.conn = None
-        output = response.read()
-        if method == "GET" and response.status == 404:
-            return ""
-        elif response.status == 304: # not modified
-            if return_headers_on_unmodified:
-                return response.getheaders()
-            else:
+        heads = { }
+        for (header, value) in response.getheaders():
+            heads[header.lower()] = value
+        return {
+            "status": response.status,
+            "reason": response.reason,
+            "headers": heads,
+            "data": response.read()
+        }
+
+    def cache(self, resource, response=None):
+        path = os.path.join(self.cache_directory, urllib.quote(resource, safe=''))
+        if response is None:
+            if not os.path.exists(path):
                 return None
-        elif response.status < 200 or response.status >= 300:
-            sys.stderr.write(output)
-            raise Exception("GitHub API problem: {0}".format(response.reason or response.status))
-        return output
+            with open(path, 'r') as fp:
+                return json.load(fp)
+        else:
+            with open(path, 'w') as fp:
+                json.dump(response, fp)
 
     def get(self, resource):
-        output = self.request("GET", resource)
-        if not output:
+        headers = { }
+        cached = self.cache(resource)
+        if cached:
```

Comment:
we can probably get by without checking all that here, since that's just the way we expect the cache to work